### PR TITLE
[Android][Feature][57643] Initialize API

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets
- [#Feature 57643: Content](https://edu-redmine.sun-asterisk.vn/issues/57643)
## WHAT
- Initialize api endpoint, network call using executor, json parsing
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/vn/ztech/software/projectgutenberg/screen/home/HomeFragment.kt [hunk 1]
```diff
@@ -31,11 +30,12 @@ class HomeFragment : BaseFragment<FragmentHomeBinding>(FragmentHomeBinding::infl
     }
 
     override fun onGetBooksSuccess(books: List<Book>) {
-        context?.toast(mutableListOf<Book>(Book("War and peace")).toString())
+        Log.d(this::class.java.name, "get book success:$books")
+        binding?.tvTest?.text = books[0].toString()
```
Comment (hunggld) [inline app/src/main/java/vn/ztech/software/projectgutenberg/screen/home/HomeFragment.kt:34]: remove log
 binding?.tvTest?.text = books.first().toString()

File: app/src/main/java/vn/ztech/software/projectgutenberg/screen/home/HomeFragment.kt [hunk 2]
```diff
@@ -31,11 +30,12 @@ class HomeFragment : BaseFragment<FragmentHomeBinding>(FragmentHomeBinding::infl
...
     }
 
     override fun onError(e: Exception?) {
-        //TODO show error
+        Log.d(this::class.java.name, "Error:$e")
```
Comment (hunggld) [inline app/src/main/java/vn/ztech/software/projectgutenberg/screen/home/HomeFragment.kt:38]: remove

File: app/src/main/java/vn/ztech/software/projectgutenberg/screen/home/HomePresenter.kt [hunk 3]
```diff
@@ -1,19 +1,22 @@
...
 ) : HomeContract.Presenter {
     private var mView: HomeContract.View? = null
 
     override fun getBooks() {
-        bookRepository.getBooks(object : OnResultListener<List<Book>> {
-            override fun onSuccess(data: List<Book>) {
-                mView?.onGetBooksSuccess(data)
+        // This page variable is for testing
+        // delete this variable or replace with a proper implementation later
+        val page = 2
```
Comment (hunggld) [inline app/src/main/java/vn/ztech/software/projectgutenberg/screen/home/HomePresenter.kt:16]: cho vào const nha anh

**Review by hungld-1302 — APPROVED** [review]

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/remote/utils/ParseDataWithJson.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            val previous = jsonObject?.getString(BaseAPIResponseEntry.PREVIOUS)
+            val resultsJsonArray = jsonObject?.getJSONArray(BaseAPIResponseEntry.RESULTS)
+            val resultObjects = resultsJsonArray?.getObjectArray<T>(keyEntity)
+            resultObjects?.let {
+                return BaseAPIResponse(
+                    count = count,
+                    next = next,
+                    previous = previous,
+                    results = it
+                )
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/remote/utils/ParseDataWithJson.kt]:
```suggestion
                return BaseAPIResponse(
                    count,
                    next,
                    previous,
                    results = it
                )
```

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/remote/utils/NetworkCall.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                )
+                mHandler.post {
+                    data?.let { callBack.onSuccess(it) }
+                }
+            } catch (e: JSONException) {
+                callBack.onError(e)
+            } catch (e: IOException) {
+                callBack.onError(e)
+            }
+        }
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/remote/utils/NetworkCall.kt:46]:
```suggestion
        }
        mExecutor.shutdow() // giải phóng mExecutor
```

File: app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -2,7 +2,7 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
     package="vn.ztech.software.projectgutenberg">
-
+    <uses-permission android:name="android.permission.INTERNET"/>
```
Comment (vietnh-1109) [inline app/src/main/AndroidManifest.xml:6]: thêm network state nữa để check internet a ơi , nó truy cập vào connectivityManager

File: app/src/main/java/vn/ztech/software/projectgutenberg/screen/home/HomeFragment.kt [hunk 7]
```diff
@@ -1,14 +1,13 @@
 package vn.ztech.software.projectgutenberg.screen.home
 
+import android.util.Log
```
Comment (minhnn-1279) [inline app/src/main/java/vn/ztech/software/projectgutenberg/screen/home/HomeFragment.kt]: remove log nhé anh :v

**Review by minhnn-1279 — APPROVED** [review]

**Review by hungvv-1301 — APPROVED** [review]

**Pull request merged**